content/language/ruby/develop.md

```diff
@@ -0,0 +1,502 @@
+
+The following is the updated `compose.yaml` file.
+
+```yaml {hl_lines="10-30"}
+version: '3'
+services:
+  web:
+    build: .
+    command: bundle exec rails s -b '0.0.0.0'
+    volumes:
+      - .:/myapp
+    ports:
+      - "3000:3000"
+    depends_on:
+      - db
+    secrets:
+      - db-password
+    environment:
+      - POSTGRES_PASSWORD_FILE=/run/secrets/db-password
+      - RAILS_ENV=test
+  db:
+    image: postgres:latest
+    secrets:
+      - db-password
+    environment:
+      - POSTGRES_PASSWORD_FILE=/run/secrets/db-password
+    volumes:
+      - postgres_data:/var/lib/postgresql/data
+
+volumes:
+  postgres_data:
+secrets:
+  db-password:
+    file: db/password.txt
+```
+
+> **Note**
+>
+> To learn more about the instructions in the Compose file, see [Compose file
+> reference](/compose/compose-file/).
+
+Before you run the application using Compose, notice that this Compose file specifies a `password.txt` file to hold the database's password. You must create this file as it's not included in the source repository.
+
+In the cloned repository's directory, create a new directory named `db` and inside that directory create a file named `password.txt` that contains the password for the database. Using your favorite IDE or text editor, add the following contents to the `password.txt` file.
+
+```text
+mysecretpassword
+```
+
+Save and close the `password.txt` file.
+
+You should now have the following contents in your `docker-ruby-on-rails`
+directory.
+
+```text
+├── docker-ruby-on-rails/
+├── app
+├── bin
+├── config
+│── db/
+│ │ └─ password.txt
+├── lib
+├── log
+├── public
+├── storage
+├── test
+├── tmp
+├── vendor
+│ ├── .dockerignore
+│ ├── .gitignore
+│ ├── confi.ru
+│ ├── Gemfile
+│ ├── Gemfile.lock
+│ ├── compose.yaml
+│ ├── Rakefile
+│ ├── Dockerfile
+│ └── README.md
+```
+
+Now, run the following `docker compose up` command to start your application.
+
+```console
+$ docker compose up --build
+```
+
+Refresh http://localhost:3000 in your browser and verify that the Whale items persisted, even after the containers were removed and ran again.
+
+Press `ctrl+c` in the terminal to stop your application.
+
+## Automatically update services
+
+Use Compose Watch to automatically update your running Compose services as you
+edit and save your code. For more details about Compose Watch, see [Use Compose
+Watch](../../compose/file-watch.md).
+
+Open your `compose.yaml` file in an IDE or text editor and then add the Compose
+Watch instructions. The following is the updated `compose.yaml` file.
+
+```yaml {hl_lines="17-2 0"}
+version: '3'
+services:
+  web:
+    build: .
+    command: bundle exec rails s -b '0.0.0.0'
+    volumes:
+      - .:/myapp
+    ports:
+      - "3000:3000"
+    depends_on:
+      - db
+    secrets:
+      - db-password
+    environment:
+      - POSTGRES_PASSWORD_FILE=/run/secrets/db-password
+      - RAILS_ENV=test
```

Comment:
This does not work, because the code expects the env var `POSTGRES_PASSWORD` to be set. We could use `env_file` and put the database password in a dotenv, e.g.

```suggestion
    environment:
      - RAILS_ENV=test
    env_file: .env
```

But it seems that there's a call missing in the example repository to set up/migrate the database

```text
db-1   | 2024-07-17 07:01:06.501 UTC [33] ERROR:  relation "whales" does not exist at character 24
db-1   | 2024-07-17 07:01:06.501 UTC [33] STATEMENT:  SELECT "whales".* FROM "whales"
web-1  | 2024-07-17 07:01:06 +0000 Rack app ("GET /" - (192.168.65.1)): #<ActionView::Template::Error:"PG::UndefinedTable: ERROR:  relation \"whales\" does not exist\nLINE 1: SELECT \"whales\".* FROM \"whales\"\n                               ^\n">
```

Reply:
In order to run the app using postgres db, you need to create a file named password.txt inside `db` folder. I am using it like a secret. But I don't see any problem to use env_file instead of secrets files.

![image](https://github.com/user-attachments/assets/eef03de2-7599-4746-a4ca-590b23b21af4)

Reply:
Yes but you also need to pass the secret to the web service, and it expects a different environment variable that holds the secret value, not a variable pointing to the filepath

Reply:
![image](https://github.com/user-attachments/assets/b546e002-13b5-4c1a-b162-d284c019f414)

The web service is obtaining the environment variable thanks to the docker secrets, otherwise we would see a connection error.  The problem is that I forgot to mention that Ruby on Rails needs to run a migration command `rake db:migrate RAILS_ENV=test` to create the schema in the database, it is the reason we get the error `"whales" does not exist `because the table has not been created. I updated the documents and the command to run is:

```
docker exec -it docker-ruby-on-rails-web-1 rake db:migrate RAILS_ENV=test
```

After that you can go to the browser and you will see:

![image](https://github.com/user-attachments/assets/6575a12f-e439-4916-a07f-6385afcc1205)

Reply:
@dvdksn  you are right, the secret is only working for the database and not for the app. I added the `env_file` to the docker compose, thanks for the suggestion.